File: ProductExtractor/models/product.py
from models.builder import Builder, MissingAttributeException
from models.category import Category


class Product:
    def __init__(
        self,
        reference: str,
        name: str,
        image: str,
        info: dict[str, str],
        category: Category,
    ):
        self.reference = reference
        self.name = name
        self.image = image
        self.category = category
        self.info = info

    def __repr__(self) -> str:
        res = "Product(\n"
        for key, val in self.__dict__.items():
            res += f"\t{key} = {str(val)}\n"
        res += ")"
        return res


class ProductBuilder(Builder[Product]):
    def __init__(self, reference: str, name: str):
        self.reference = reference
        self.name = name
        self.image = None
        self.category = None
        self.info = None

    def __repr__(self) -> str:
        res = "Product(\n"
        for key, val in self.__dict__.items():
            res += f"\t{key} = {str(val)}\n"
        res += ")"
        return res

    def set_image(self, image: str):
        self.image = image

    def set_category(self, category: Category):
        self.category = category

    def set_info(self, info: dict[str, str]):
        self.info = info

    def build(self) -> Product:
        if self.image is None:
            raise MissingAttributeException(
                "Attempted to construct Product without setting image attribute"
            )

        if self.info is None:
            raise MissingAttributeException(
                "Attempted to construct Product without setting info attribute"
            )

        if self.category is None:
            raise MissingAttributeException(
                "Attempted to construct Product without setting category attribute"
            )

        return Product(
            self.reference,
            self.name,
            self.image,
            self.info,
            self.category,
        )

File: offer-extractor/models/product.ts
import { Category } from "./category.ts";

export class Product {
    reference: string;
    name: string;
    image: string;
    info: Map<string, string>;
    category: Category;

    constructor(
        reference: string,
        name: string,
        image: string,
        info: Map<string, string>,
        category: Category,
    ) {
        this.reference = reference;
        this.name = name;
        this.image = image;
        this.category = category;
        this.info = info;
    }

    toString(): string {
        let res = "Product(\n"
        Object.entries(this).forEach((key, val) => {
            res += `\t${key} = ${val}\n`;
        })
        res += ")"
        return res
    }
}

export class ProductBuilder {
    reference: string;
    name: string;
    image: string | null;
    info: Map<string, string> | null;
    category: Category | null;

    constructor(reference: string, name: string) {
        this.reference = reference;
        this.name = name;
        this.image = null;
        this.category = null;
        this.info = null;
    }


    setImage(image: string) {
        this.image = image;
    }

    setCategory(category: Category) {
        this.category = category
    }

    setInfo(info: Map<string, string>) {
        this.info = info;
    }

    build(): Product {
        if (this.image === null) {
            throw Error(
                "Attempted to construct Product without setting image attribute"
            );
        }

        if (this.info === null) {
            throw Error(
                "Attempted to construct Product without setting info attribute"
            );
        }

        if (this.category === null) {
            throw Error(
                "Attempted to construct Product without setting category attribute"
            );
        }

        return new Product(
            this.reference,
            this.name,
            this.image,
            this.info,
            this.category
        );
    }
}
